File: 530-minimum-absolute-difference-in-bst/530-minimum-absolute-difference-in-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var getMinimumDifference = function(root) {
    let res = Infinity;
    
    const walk = (node) => {
        if(node === null){
          return [Infinity, -Infinity];
        }
        
        const [minleft, maxleft] = walk(node.left);
        const [minright, maxright] = walk(node.right);
        
        res = Math.min(res, 
                       Math.abs(maxleft - node.val),
                       Math.abs(minright - node.val));
        
        return [Math.min(node.val, minleft), Math.max(node.val, maxright)];
        
    };
    
    walk(root);
    return res
};

File: Project Phase Camp/530-minimum-absolute-difference-in-bst/530-minimum-absolute-difference-in-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        
        self.min = float("inf")
        
        def dfs(node):
            if not node:
                return float("-inf"),float("inf")
            
#             i recieve left maximum and right minimum from the function call
#              l_max and r_min

            l_max,l_min = dfs(node.left)
            r_max,r_min = dfs(node.right)
            
            self.min = min(abs(node.val - l_max), self.min, abs(node.val - r_min))
            
            mx = max(l_max, node.val,r_max)
            mn = min(node.val, l_min, r_min)
            
            return mx,mn
            
        dfs(root)
        return self.min